### PR TITLE
server: Add authentication using package mode

### LINKED ISSUE #16: Implement authentication for package data
## Feature Request

Signing each piece of data can be inefficient, because the server often requires more than one piece of data.
However, if the raw value is concated and signed as it is, the client may be fooled by the authenticated value.

`Sign('abc' | 'd') == Sign('ab' | 'cd')`

Therefore, after encoding using `btoa`, etc., characters such as `,` should be added between values to prevent manipulation of the internal values.

https://github.com/0-Auth/0Auth/blob/b85bfec96e044991414de6100dd0495613e269da/packages/server/src/index.ts#L19-L23
https://github.com/0-Auth/0Auth/blob/b85bfec96e044991414de6100dd0495613e269da/packages/server/src/index.ts#L31-L35

**Is your feature request related to a problem? Please describe.**  

Signing and verifying signatures are more expensive than common tasks.

**Describe the solution you'd like**  

If the values are used together, they are bundled and signed.

**Describe alternatives you've considered**  

**Additional context**

### PR DESCRIPTION
**Pull request by HyeockJinKim**
# Summary 

Implement authentication and verification functions to check values
used together at once

- It closes #16 

## How to Review? 

- Please refer to my self-review submitted as the review.

### CODE REVIEW

File: packages/server/src/index.ts [hunk 1]
```diff
@@ -17,13 +24,13 @@ export function authPrivacy(properties: Property[], secret: SecretKey): Signatur
...
-// TODO: @ts-ignore should be removed, when below function is implemented.
-// @ts-ignore
 export function authPackage(
   properties: Property[],
   secret: SecretKey
 ): Signature {
-  // TODO: Not implemented yet.
+  const encodings = properties.map(property => utf8ToBase64(property.key)+":"+utf8ToBase64(property.value)),
+    hashValue = hash(encodings.join(','));
+  return signAccordingToKeyType(hashValue, secret.key, secret.type);
```
Comment (HyeockJinKim) [inline packages/server/src/index.ts:33]: After simply processing base64 encoding, it is hashed and signed.

File: packages/server/src/index.ts [hunk 2]
```diff
@@ -35,12 +42,12 @@ export function verifyPrivacy(properties: Property[], sign: Signature, publicKey
...
-// @ts-ignore
 export function verifyPackage(
   properties: Property[],
   sign: Signature,
   publicKey: PublicKey
 ): boolean {
-  // TODO: Not implemented yet.
+  const encodings = properties.map(property => utf8ToBase64(property.key)+":"+utf8ToBase64(property.value)),
+    hashValue = hash(encodings.join(','));
+  return verifyAccordingToKeyType(hashValue, sign.value, publicKey.key, publicKey.type);
```
Comment (HyeockJinKim) [inline packages/server/src/index.ts:52]: After simply processing base64 encoding, the signature is verified through the hashed value.

File: packages/server/test/auth.spec.ts [hunk 3]
```diff
@@ -129,6 +129,24 @@ describe("test privacy mode authentication", () => {
...
 
+    expect(verifyPackage(properties, sign, { key: key.getPublic('hex'), type: KeyType.ECDSA }))
+  });
+  it('test EdDSA authentication & verification using package mode', function() {
+    const eddsa = new EdDSA('ed25519');
+    const key = eddsa.keyFromSecret('2ef40452ec154cd38efdc8ffa52e7f513f7d2b2a77e028342bde96c369e4f77a');
+    const sign = authPackage(properties, {key: key.getSecret('hex'), type: KeyType.EDDSA});
+
+    expect(verifyPackage(properties, sign, { key: key.getPublic('hex'), type: KeyType.EDDSA }))
+  });
```
Comment (HyeockJinKim) [inline packages/server/test/auth.spec.ts:151]: Test code for authentication and verification of package mode.

File: packages/server/test/auth.spec.ts [hunk 4]
```diff
@@ -1,134 +1,156 @@
...
+    const key = ecdsa.keyFromPrivate('2ef40452ec154cd38efdc8ffa52e7f513f7d2b2a77e028342bde96c369e4f77a');
+    const sign = authPrivacy(properties, { key: key.getPrivate('hex'), type: KeyType.ECDSA });
 
-    expect(verifyPrivacy(properties, sign, {key: key.getPublic("hex"), type: KeyType.ECDSA})).to.be.equal(true);
-    expect(verifyPrivacy(properties2, sign, {key: key.getPublic("hex"), type: KeyType.ECDSA})).to.be.equal(true);
+    expect(verifyPrivacy(properties, sign, { key: key.getPublic('hex'), type: KeyType.ECDSA })).to.be.equal(true);
+    expect(verifyPrivacy(properties2, sign, { key: key.getPublic('hex'), type: KeyType.ECDSA })).to.be.equal(true);
   });
-})
+});
```
Comment (anthonyminyungi) [inline packages/server/test/auth.spec.ts:134]: I think it's just formatted from L1 to here.

**Pull request merged**